### ISSUE TITLE
KVCompareAndSet should not ignore ExpireAt

### ISSUE THREAD

**Issue by lieut-data** [mattermost/mattermost-server#13656]
`KVCompareAndSet` doesn't test `ExpireAt`, allowing it to compare against existing, stale values when deciding whether to write to the row. Change the implementation to consider `ExpireAt`, such that an expired row is only successfully written if the old value is given as `nil`, effectively treating it the same as if the row had never existed.

Discovered during investigation into [https://mattermost.atlassian.net/browse/MM-21328](https://mattermost.atlassian.net/browse/MM-21328).

----

If you're interested please comment here and come [join our "Contributors" community channel](https://community.mattermost.com/core/channels/tickets) on our daily build server, where you can discuss questions with community members and the Mattermost core team. For technical advice or questions, please [join our "Developers" community channel](https://community.mattermost.com/core/channels/developers).

New contributors please see our [Developer's Guide](https://developers.mattermost.com/contribute/getting-started/).

JIRA: https://mattermost.atlassian.net/browse/MM-21672

**Comment by 0xlilyhoffman** [comment]
Looks interesting!

Hmm...
If the goal is to treat expired rows as if they don’t exist, would it make sense to simply delete all expired values before performing the CompareAndSet? 
Is there a reason to save the expired records?

**Comment by lieut-data** [comment]
@0xlilyhoffman, we do indeed have a background process that purges such records, but it intentionally runs at a slow cadence. Deleting the records prior to a `KVCompareAndSet` might require a full table scan at present, which could impact performance. It might also allow for a race condition where a new, and very quickly expiring record gets written after the purge but before the `CompareAndSet`.

I /think/ it should be straightforward to handle this case by tweaking the SQL query with an additional `WHERE` clause of `(ExpireAt = 0 OR ExpireAt > :CurrentTime)` and as used elsewhere in the store. The majority of the work would be adding unit tests to verify the behaviour.

**Comment by igomonov88** [comment]
@lieut-data I cant try to do it

**Comment by 0xlilyhoffman** [comment]
Ah I see. 
Thanks @lieut-data for following up. 

I’d be interested in taking this, if that’s ok. 
Is this time sensitive in any way?

**Comment by lieut-data** [comment]
@0xlilyhoffman, not time-sensitive -- thanks to both you and @igomonov88 for volunteering to help here! Did you want to coordinate efforts, or decide who might take the lead on this effort?

**Comment by igomonov88** [comment]
@0xlilyhoffman hello, nice to meet you. Let's make a sync since the task was assigned to both of us. Maybe you have any preferences for this task? If you want, you can fix the implementation of the method and I can add new unit tests?

**Comment by 0xlilyhoffman** [comment]
@igomonov88 Hello! Thanks for reaching out - I wasn't sure how you wanted to go about this. Yes, that could work. I started working on it and am actually having a few issues/confusions (with both implementation and testing). I'll push my WIP to my own forked repo (https://github.com/0xlilyhoffman/mattermost-server), and maybe we can discuss details from there. I'd love to hear your feedback / approach.

**Comment by 0xlilyhoffman** [comment]
@lieut-data regarding deleting entries before the KVCompareAndSet: would invoking some “DeleteIfExpired(pluginId, key string)” to only delete the key in question just before updating the entry in the database solve this issue in a safe and performant way?

For reference, I was thinking something like this: https://github.com/0xlilyhoffman/mattermost-server/commit/6c8459d393fa7d04f981715a4cab9b5afe399596

**Comment by lieut-data** [comment]
@0xlilyhoffman, I like your approach! There's a potential that a row is created and immediately expires after you try to delete, but this would be very unusual, and still return correct results depending on your frame of reference relative to the wall clock.

I'd suggest not exporting `DeleteIfExpired` (and thus not writing an explicit unit test for same), since we probably don't need to share that method with a client of the store, and we can test the semantics in context of `KVCompareAndSet` instead.

**Comment by 0xlilyhoffman** [comment]
Thanks @lieut-data and @igomonov88 for your feedback! 
I really appreciate it.

I submitted a draft PR (https://github.com/mattermost/mattermost-server/pull/13757) so that Mattermost CI servers would run all tests. All of the tests pass, but it’s failing on “mattermost/serverless-ops/github - This PR needs to sync and merge”

What are the recommended steps to addressing “This PR needs to sync and merge”?

For some context:
As I was working on my branch MM-21672, it fell behind master. I just updated my forked repo so that it's mater branch is in sync with the master branch of mattermost/mattermost-server. Would this issue be solved by merging my own (now in sync) master into my MM-21672 branch and pushing MM-21672 to my forked repo? Or can this be solved later in the PR process by a “rebase and merge”.

Also - apologies if this should be straightforward. This is my first time contributing to open source!
